From 2a1603aae01e161e6fdda1e1effb4c82490e0a4e Mon Sep 17 00:00:00 2001
From: PyScotch <pyscotch@noreply.invalid>
Date: Tue, 13 Oct 2026 13:10:36 +0000
Subject: [PATCH 013/240] Bugfix: set default file names in `dummysizes`
 [report R. Perry]

Run with fewer than two file names, `dummysizes` dereferenced the unset
`nameptr` fields of its file array (reported by UBSan and a crash).
Set the default file names to "-", as the other command-line tools do
through `fileBlockInit()`.

Also test the last slot that `fgets()` actually fills when checking for
an over-long line, and refer to the existing `File` fields in the
`SCOTCH_DEBUG_MAIN1` block, which did not compile.

Refs: COR-20
---
 src/libscotch/dummysizes.c | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/src/libscotch/dummysizes.c b/src/libscotch/dummysizes.c
index 0bad641..c45eb2f 100644
--- a/src/libscotch/dummysizes.c
+++ b/src/libscotch/dummysizes.c
@@ -1,4 +1,4 @@
-/* Copyright 2004,2007-2010,2012,2014,2018,2019,2021,2023-2025 IPB, Universite de Bordeaux, INRIA & CNRS
+/* Copyright 2004,2007-2010,2012,2014,2018,2019,2021,2023-2026 IPB, Universite de Bordeaux, INRIA & CNRS
 **
 ** This file is part of the Scotch software package for static mapping,
 ** graph partitioning and sparse matrix ordering.
@@ -56,7 +56,7 @@
 /**                # Version 6.1  : from : 09 feb 2021     **/
 /**                                 to   : 22 jun 2021     **/
 /**                # Version 7.0  : from : 25 aug 2019     **/
-/**                                 to   : 13 aug 2025     **/
+/**                                 to   : 13 oct 2026     **/
 /**                                                        **/
 /************************************************************/
 
@@ -191,8 +191,10 @@ char *                      argv[])
 
   suffptr = "";                                   /* No suffix */
 
-  for (i = 0; i < C_FILENBR; i ++)                /* Set default stream pointers */
+  for (i = 0; i < C_FILENBR; i ++) {              /* Set default stream pointers and names */
+    C_fileTab[i].nameptr = "-";
     C_fileTab[i].fileptr = ((C_fileTab[i].flagval & FILEMODE) == FILEMODER) ? stdin : stdout;
+  }
   for (i = 1; i < argc; i ++) {                   /* Loop for all option codes */
     if ((argv[i][0] != '+') &&                    /* If found a file name      */
         ((argv[i][0] != '-') || (argv[i][1] == '\0'))) {
@@ -363,7 +365,7 @@ char *                      argv[])
     int                 subsnum;
 
     if ((strlen (chartab)) >= (CHARMAX - 1) &&    /* If line read is at least as long as maximum size              */
-        (chartab[CHARMAX - 1] != '\n')) {         /* And last character is not a newline, that is, some is missing */
+        (chartab[CHARMAX - 2] != '\n')) {         /* And last character is not a newline, that is, some is missing */
       fprintf (stderr, "dummysizes: ERROR: line too long\n");
       exit    (EXIT_FAILURE);
     }
@@ -405,9 +407,9 @@ char *                      argv[])
 
 #ifdef SCOTCH_DEBUG_MAIN1
   for (i = 0; i < C_FILENBR; i ++) {              /* For all file names     */
-    if ((C_fileTab[i].name[0] != '-') ||          /* If not standard stream */
-        (C_fileTab[i].name[1] != '\0')) {
-      fclose (C_fileTab[i].pntr);                 /* Close the stream */
+    if ((C_fileTab[i].nameptr[0] != '-') ||       /* If not standard stream */
+        (C_fileTab[i].nameptr[1] != '\0')) {
+      fclose (C_fileTab[i].fileptr);              /* Close the stream */
     }
   }
 #endif /* SCOTCH_DEBUG_MAIN1 */
-- 
2.47.3

